scipy/sparse/linalg/_isolve/symmlq.py

```diff
@@ -0,0 +1,239 @@
+
+        .. warning::
+
+           The default value for `atol` will be changed in a future release.
+           For future compatibility, specify `atol` explicitly.
+    haptol : float, optional
+        Tolerance for happy breakdown, ``abs(np.inner(r.conjugate(), Mr)) <
+        haptol``.
+        Default is 1.0e-18.
+    maxiter : int, optional
+        Maximum number of iterations.  Iteration will stop after maxiter
+        steps even if the specified tolerance has not been achieved.
+        Default is ``min(10000, n * 10)``, where ``n = A.shape[0]``.
+    M : {sparse matrix, ndarray, LinearOperator}
+        Inverse of the preconditioner of A. M should approximate the
+        inverse of A and be easy to solve for (see Notes). Effective
+        preconditioning dramatically improves the rate of convergence,
+        which implies that fewer iterations are needed to reach a given
+        error tolerance. By default, no preconditioner is used.
+    callback : function, optional
+        User-supplied function to call after each iteration. It is called
+        as `callback(rnorm)`, where `rnorm` is the current preconditioned
+        residual norm.
+    show : bool, optional
+        Specify ``show = True`` to show the convergence, ``show = False`` is
+        to close the output of the convergence.
+        Default is `False`.
+
+    Returns
+    -------
+    x : ndarray
+        The converged solution.
+    info : int
+        Provides convergence information:
+
+            - 0  : successful exit
+            - >0 : convergence to tolerance not achieved, number of iterations
+            - <0 : illegal input or breakdown
+
+    Notes
+    -----
+    The Symmetric LQ algorithm (SYMMLQ) is a variant of CG method, which is
+    designed to avoid LU decomposition and breakdown, and can be applied to
+    symmetric indefinite systems. Unlike MINRES, SYMMLQ keeps the residual
+    orthogonal to all previous ones (Don't minimize anything).
+
+    References
+    ----------
+    .. [1] C. C. Paige, M. A. Saunders, Solution of Sparse Indefinite Systems
+           of Linear Equations, SIAM J. Numer. Anal., Vol. 12, No. 4, 1975.
+
+    This file is a python translation of PETSc implementation:
+        https://gitlab.com/petsc/petsc/-/blob/main/src/ksp/ksp/impls/symmlq/symmlq.c
+
+    Examples
+    --------
+    >>> from scipy.sparse import csc_matrix
+    >>> from scipy.sparse.linalg import symmlq
+    >>> A = csc_matrix([[2, -1, 0], [-1, 0.25, -1], [0, -1, 0]], dtype=float)
+    >>> b = np.array([1, 2, 3], dtype=float)
+    >>> x, exitCode = symmlq(A, b)
+    >>> print(exitCode)            # 0 indicates successful convergence
+    0
+    >>> np.allclose(A.dot(x), b)
+    True
+    """
+
+    A, M, x, b, postprocess = make_system(A, M, x0, b)
+
+    n = A.shape[0]
+    if maxiter is None:
+        maxiter = min(10000, n * 10)
+
+    # Check if the R.H.S is a zero vector
+    bnorm = np.linalg.norm(b)
```

Comment:
> Yes but what is the point of having sparse algorithms that can't take sparse inputs. 

In most PDEs equations and its discrete methods, dense right-hand side is usually used and depends on external force in physical equations, discrete method and required RHS/resiual/solution form in Krylov iteration. In the so-called sparse linear iterative algorithm (stationary iteration & Krylov iteration), the coefficient matrix is sparse, the right-hand side is dense. It is also this case with the well-known SpMV (sparse matrix-vector multiplication). Unless you have specific type of problems that resulit in sparse vectors that remain sparse during the linear solution process, it's more efficient to use dense right-hand side.
In my impression, some sparse direct solvers can use sparse right-hand side such as MUMPS/UMFPACK/SuperLU, but for sparse iterative solvers, we don't usually do that because the sparsity of vectors in the iteration cannot be remained.